badrcptto.c

```diff
@@ -0,0 +1,67 @@
+#include <unistd.h>
+#include "byte.h"
+#include "constmap.h"
+#include "control.h"
+#include "env.h"
+#include "fmt.h"
+#include "str.h"
+#include "stralloc.h"
+#include "strerr.h"
+
+extern void die_control();
+extern void die_nomem();
+
+static void _badrcptto_log_rejection(char *recipient)
+{
+  char smtpdpid[32];
+  char *remoteip = env_get("TCPREMOTEIP");
+  if (!remoteip) remoteip = "unknown";
+  str_copy(smtpdpid + fmt_ulong(smtpdpid,getppid())," ");
+  strerr_warn5("rcptcheck: badrcptto ",smtpdpid,remoteip," ",recipient,0);
+}
+
+static int _badrcptto_reject_exact_address(struct constmap map, stralloc address)
+{
+  return (1 && constmap(&map,address.s,address.len - 1));
+}
+
+static int _badrcptto_reject_whole_domain(struct constmap map, stralloc address)
+{
+  /* why not just comment out the domain in control/rcpthosts? */
+  int j = byte_rchr(address.s,address.len,'@');
+  return ((j < address.len) && (constmap(&map,address.s + j,address.len - j - 1)));
+}
+
+static int _badrcptto_reject_string(char *string)
+{
+  stralloc addr = {0};
+  stralloc brt = {0};
```

Comment:
I think I missed when those two are set `stralloc_free()`. Or does it not matter since it is a short-lived program?

I realize that there is not even any [stralloc_free()](https://skarnet.org/software/skalibs/libstddjb/stralloc.html) in DJB code. It could be introduced later.

Reply:
> The benefits of using stralloc are as follows: [...]
> - If every heap-allocated object is represented by a stralloc, then it is very easy to identify what pointer is in the heap. When you stop using a pointer p, should you free it ? Sometimes it's not easy to find out. When you stop using a stralloc sa, you know you must call stralloc_free(&sa). Store your strong references as strallocs and weak references as simple pointers, and never free simple pointers. This policy allows me to boast that no skarnet.org software has ever leaked memory.  -- https://skarnet.org/software/skalibs/libstddjb/stralloc.html

Reply:
Other wraparound of DJB code do include `stralloc_free`, like fefe's https://www.fefe.de/libowfat/:

```
#include <stdlib.h>
#include "stralloc.h"

void stralloc_free(stralloc *sa) {
  if (sa->s) free(sa->s);
  sa->s=0;
  sa->a=sa->len=0;
}
```